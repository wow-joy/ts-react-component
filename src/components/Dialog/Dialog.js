import React, { PureComponent } from "react";
import PropTypes from "prop-types";
import { Type1 as Btn_1, Type3 as Btn_3 } from "../Btn";
import styled from "styled-components";

const Wrap = styled.div`
  width: 560px;
  position: relative;
  background: #fff;
  box-shadow: 0 0 4px 0 rgba(0, 0, 0, 0.2);
  padding-bottom: 24px;
  ${props => props.defaultStyles};
`;

const CloseBtn = styled.div`
  position: absolute;
  top: 0;
  right: 0;
  font-size: 16px;
  width: 38px;
  height: 38px;
  text-align: center;
  line-height: 38px;
  color: #909090;
  cursor: pointer;
  z-index: 2;
  &::before {
    content: "";
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    margin: auto;
    right: 0;
    width: 1em;
    height: 2px;
    background: currentColor;
    transform: rotate(45deg);
  }
  &::after {
    content: "";
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    margin: auto;
    right: 0;
    width: 1em;
    height: 2px;
    background: currentColor;
    transform: rotate(-45deg);
  }
`;
const HeaderDom = styled.div`
  width: 100%;
  height: 38px;
  line-height: 38px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  padding: 0 10px;
  position: relative;
  &::after {
    content: "";
    height: 1px;
    background: #ebebeb;
    position: absolute;
    bottom: 0;
    left: 10px;
    right: 10px;
    transform: scaleY(0.5);
    transform-origin: 50% 100%;
  }
`;
const Content = styled.section`
  padding: 24px 20px;
`;

const Btns = styled.div`
  display: block;
  text-align: center;
`;

class Dialog extends PureComponent {
  closeHandle = e => {
    const { onClose } = this.props;
    onClose && onClose(e);
  };
  clickHandle = (e, index) => {
    const { onClick } = this.props;
    if (onClick && onClick(e, index) === false) {
      return;
    }
    this.closeHandle(e);
  };
  render() {
    const {
      defaultStyles,
      className,
      header,
      headerText,
      children,
      btnsText,
      btns,
      showCloseBtn = true
    } = this.props;
    const Header =
      header && header !== true ? header : props => <HeaderDom {...props} />;

    const Btn_default = [Btn_1, Btn_3];

    return (
      <Wrap
        defaultStyles={defaultStyles}
        className={"wjc-dialog " + (className || "")}
      >
        {showCloseBtn && (
          <CloseBtn
            className={"wjc-dialog-btn__close"}
            onClick={this.closeHandle}
          />
        )}
        {header !== false && (
          <Header className={"wjc-dialog-header"}>{headerText}</Header>
        )}

        <Content className={"wjc-dialog-content"}>{children}</Content>
        <Btns className={"wjc-dialog-btns"}>
          {(btnsText || ["提交", "取消"]).map((ele, index) => {
            const Btn = (btns && btns[index]) || Btn_default[index] || Btn_1;
            return (
              <Btn
                key={index}
                onClick={e => this.clickHandle(e, index)}
                defaultStyles={`margin: 0 5px;min-width: 130px;`}
              >
                {ele}
              </Btn>
            );
          })}
        </Btns>
      </Wrap>
    );
  }
}

Dialog.propTypes = {
  defaultStyles: PropTypes.string,
  className: PropTypes.string,
  header: PropTypes.any,
  headerText: PropTypes.string,
  children: PropTypes.node,
  btnsText: PropTypes.array,
  btns: PropTypes.array,
  onClick: PropTypes.func,
  onClose: PropTypes.func,
  showCloseBtn: PropTypes.bool
};
export default Dialog;
