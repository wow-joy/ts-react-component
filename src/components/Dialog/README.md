# Dialog

对话框组件

_建议结合 `Pop` 组件使用_ [https://github.com/wow-joy/wowjoy-component/tree/master/src/components/Pop]

_也可结合 `pop`工具使用_ [https://github.com/wow-joy/wowjoy-component/tree/master/src/tools/pop]

## 使用

- 基础用法

```
import React from 'react'
import { Dialog, Dialog_alert } from 'wowjoy-component'

const Foo = () => <Dialog> 这是一个对话框 </Dialog>
const Foo1 = () => <Dialog_alert onClick={()=>alert(1)} > 这是一个提示框 </Dialog_alert>
```

## APIs

| 属性          | 子属性 | 描述                                                                                                                                                                                                                         |      类型      |      默认值      |
| ------------- | ------ | :--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | :------------: | :--------------: |
| className     |        | 顶层样式 class                                                                                                                                                                                                               |     string     |                  |
| defaultStyles |        | 顶层默认样式                                                                                                                                                                                                                 |     string     |                  |
| header        |        | 标题组件渲染函数: 例如 ()=><Foo\><\/Foo> ;<br/> 当值为`false`时不显示标题;为空时显示默认标题                                                                                                                                 | func \|\| bool |                  |
| headerText    |        | 标题文本                                                                                                                                                                                                                     |     string     |                  |
| children      |        | 组件包裹内容                                                                                                                                                                                                                 |      node      |                  |
| btns          |        | 按钮节点数组, 请在数组中传递 function 而不是已经实例化的组件:<br/> 例如 [ props =><Foo {...props}>{props.children}<\/Foo>] props 可以拿到默认设定的`onClick`, `className`<br/>与`Btn`组件结合使用，可以直接传入[Btn_1,Btn_2] |     array      |                  |
| btnsText      |        | 按钮文案数组: 数组个数控制按钮数量                                                                                                                                                                                           |     array      | ["提交", "取消"] |
| onClick       |        | 点击事件。包含组件内除关闭按钮外的所有按钮。 接受两个参数`(event,index)`<br/> `event`: 点击事件<br/>`index`: 被点击按钮的下标<br/> `return false`可以阻止组件被关闭                                                          |      func      |                  |
| onClose       |        | 关闭事件，点击`x`按钮触发，`onClick`没有`return false` 也会触发                                                                                                                                                              |      func      |                  |
| showCloseBtn  |        | 是否显示关闭按钮                                                                                                                                                                                                             |      bool      |       true       |

## classApi

| class                   | 描述        |
| ----------------------- | :---------- |
| wjc-dialog-btn\_\_close | 关闭按钮    |
| wjc-dialog-header       | dialog 头部 |
| wjc-dialog-content      | dialog 内容 |
| wjc-dialog-btns         | dialog 按钮 |
