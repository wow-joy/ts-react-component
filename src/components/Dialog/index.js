import Dialog from './Dialog';
import Hoc from '../../tools/Hoc';
import type1 from "./types/type1";
import confirm from "./types/confirm";
import alert from "./types/alert";

export const Type1 = Hoc(type1)(Dialog);
export const Confirm = Hoc(confirm)(Dialog);
export const Alert = Hoc(alert)(Dialog);
export default Dialog