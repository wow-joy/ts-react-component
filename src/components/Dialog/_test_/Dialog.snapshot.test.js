import React from "react";
import Dialog from "../index.js";
import renderer from "react-test-renderer";
import * as TestUtils from "react-dom/test-utils";
import { MemoryRouter, Route } from "react-router-dom";
import { mount, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });

const { renderIntoDocument } = TestUtils;

describe("Dialog", () => {
  const reactObj = <Dialog>Dialog Content</Dialog>;
  it("renders correctly", () => {
    const tree = renderer.create(reactObj).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
