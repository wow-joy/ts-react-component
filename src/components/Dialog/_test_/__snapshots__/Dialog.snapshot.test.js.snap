// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog renders correctly 1`] = `
<div
  className="wjc-dialog  sc-bwzfXH hSyxhq"
>
  <div
    className="wjc-dialog-btn__close sc-htpNat bbETrz"
    onClick={[Function]}
  />
  <div
    className="wjc-dialog-header sc-bxivhb ecavox"
  />
  <section
    className="wjc-dialog-content sc-ifAKCX beXtUg"
  >
    Dialog Content
  </section>
  <div
    className="wjc-dialog-btns sc-EHOje cuidGM"
  >
    <span
      className="wjc-btn   sc-bdVaJa LGHCx"
      onClick={[Function]}
    >
      提交
    </span>
    <span
      className="wjc-btn   sc-bdVaJa hFqIKz"
      onClick={[Function]}
    >
      取消
    </span>
  </div>
</div>
`;
